File: mobile/app/src/main/java/com/remotedesktop/input/InputEventEncoder.kt
package com.remotedesktop.input

import org.json.JSONArray
import org.json.JSONObject

object InputEventEncoder {

    const val TYPE_TEXT_MAX_LENGTH: Int = 256

    val BUTTONS: Set<String> = setOf("left", "right", "middle")
    val MODIFIERS: Set<String> = setOf("control", "shift", "alt", "meta")
    val KEYS: Set<String> = buildSet {
        addAll(('A'..'Z').map { it.toString() })
        addAll((0..9).map { it.toString() })
        addAll(listOf("Space", "Enter", "Tab", "Backspace", "Delete", "Escape"))
        addAll(listOf("Left", "Right", "Up", "Down"))
        addAll(listOf("Home", "End", "PageUp", "PageDown"))
        addAll((1..12).map { "F$it" })
    }

    private fun normalised(value: Double): Double = value.coerceIn(0.0, 1.0)

    fun mouseMove(x: Double, y: Double): String =
        JSONObject().apply {
            put("type", "mousemove")
            put("x", normalised(x))
            put("y", normalised(y))
        }.toString()

    fun mouseDown(button: String): String {
        require(button in BUTTONS) { "Unknown mouse button: $button" }
        return JSONObject().apply {
            put("type", "mousedown")
            put("button", button)
        }.toString()
    }

    fun mouseUp(button: String): String {
        require(button in BUTTONS) { "Unknown mouse button: $button" }
        return JSONObject().apply {
            put("type", "mouseup")
            put("button", button)
        }.toString()
    }

    fun click(x: Double, y: Double, button: String): String {
        require(button in BUTTONS) { "Unknown mouse button: $button" }
        return JSONObject().apply {
            put("type", "click")
            put("x", normalised(x))
            put("y", normalised(y))
            put("button", button)
        }.toString()
    }

    fun scroll(dx: Int, dy: Int): String =
        JSONObject().apply {
            put("type", "scroll")
            put("dx", dx)
            put("dy", dy)
        }.toString()

    fun keyDown(key: String, modifiers: List<String> = emptyList()): String =
        keyEvent("keydown", key, modifiers)

    fun keyUp(key: String, modifiers: List<String> = emptyList()): String =
        keyEvent("keyup", key, modifiers)

    private fun keyEvent(type: String, key: String, modifiers: List<String>): String {
        require(key in KEYS) { "Unknown key: $key" }
        for (m in modifiers) require(m in MODIFIERS) { "Unknown modifier: $m" }
        return JSONObject().apply {
            put("type", type)
            put("key", key)
            put("modifiers", JSONArray(modifiers))
        }.toString()
    }

    fun typeText(text: String): String {
        require(text.isNotEmpty()) { "text must not be empty" }
        val truncated = if (text.length > TYPE_TEXT_MAX_LENGTH) text.substring(0, TYPE_TEXT_MAX_LENGTH) else text
        return JSONObject().apply {
            put("type", "type")
            put("text", truncated)
        }.toString()
    }

    fun ping(id: Int): String {
        require(id >= 0)
        return JSONObject().apply {
            put("type", "ping")
            put("id", id)
        }.toString()
    }

    fun pong(id: Int): String {
        require(id >= 0)
        return JSONObject().apply {
            put("type", "pong")
            put("id", id)
        }.toString()
    }
}

File: mobile/ios/RemoteDesktop/Sources/InputEventEncoder.swift
import Foundation

enum InputEventEncoder {
    static let buttons: Set<String> = ["left", "right", "middle"]
    static let modifiers: Set<String> = ["control", "shift", "alt", "meta"]
    static let typeTextMaxLength = 256

    static let keys: Set<String> = {
        var set = Set<String>()
        for ch in "ABCDEFGHIJKLMNOPQRSTUVWXYZ" { set.insert(String(ch)) }
        for n in 0...9 { set.insert(String(n)) }
        set.formUnion(["Space", "Enter", "Tab", "Backspace", "Delete", "Escape"])
        set.formUnion(["Left", "Right", "Up", "Down"])
        set.formUnion(["Home", "End", "PageUp", "PageDown"])
        for n in 1...12 { set.insert("F\(n)") }
        return set
    }()

    static func mouseMove(x: Double, y: Double) -> String {
        encode(["type": "mousemove", "x": clamp(x), "y": clamp(y)])
    }

    static func mouseDown(button: String) throws -> String {
        guard buttons.contains(button) else { throw EncoderError.unknownButton(button) }
        return encode(["type": "mousedown", "button": button])
    }

    static func mouseUp(button: String) throws -> String {
        guard buttons.contains(button) else { throw EncoderError.unknownButton(button) }
        return encode(["type": "mouseup", "button": button])
    }

    static func click(x: Double, y: Double, button: String) throws -> String {
        guard buttons.contains(button) else { throw EncoderError.unknownButton(button) }
        return encode(["type": "click", "x": clamp(x), "y": clamp(y), "button": button])
    }

    static func scroll(dx: Int, dy: Int) -> String {
        encode(["type": "scroll", "dx": dx, "dy": dy])
    }

    static func keyDown(key: String, modifiers mods: [String] = []) throws -> String {
        try ensureKey(key)
        for m in mods { guard modifiers.contains(m) else { throw EncoderError.unknownModifier(m) } }
        return encode(["type": "keydown", "key": key, "modifiers": mods])
    }

    static func keyUp(key: String, modifiers mods: [String] = []) throws -> String {
        try ensureKey(key)
        for m in mods { guard modifiers.contains(m) else { throw EncoderError.unknownModifier(m) } }
        return encode(["type": "keyup", "key": key, "modifiers": mods])
    }

    static func typeText(_ text: String) throws -> String {
        guard !text.isEmpty else { throw EncoderError.emptyText }
        let truncated = String(text.prefix(typeTextMaxLength))
        return encode(["type": "type", "text": truncated])
    }

    static func ping(id: Int) -> String { encode(["type": "ping", "id": id]) }
    static func pong(id: Int) -> String { encode(["type": "pong", "id": id]) }

    private static func clamp(_ v: Double) -> Double { min(1.0, max(0.0, v)) }

    private static func ensureKey(_ key: String) throws {
        guard keys.contains(key) else { throw EncoderError.unknownKey(key) }
    }

    private static func encode(_ dict: [String: Any]) -> String {
        let data = try! JSONSerialization.data(withJSONObject: dict, options: [])
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    enum EncoderError: Error {
        case unknownButton(String)
        case unknownKey(String)
        case unknownModifier(String)
        case emptyText
    }
}
